### PR TITLE
fix traffic accounting in internal statistics / metrics

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Fix traffic accounting for HTTP/VST in internal statistics / metrics.
Outgoing VST traffic was not accounted for previously.
Additionally, the traffic was not accounted for the request that initiates a protocol change from HTTP/1.1 to HTTP/2.
Finally, the metrics histograms that summed up traffic were not correctly reporting values since 3.10.

For consistency, this PR also adds a new metrics `arangodb_http1_connections_total`, which counts the number of HTTP/1.1 connections. There were previously only metrics for HTTP/2 and VST connections.

- [x] :hankey: Bugfix
- [ ] :pizza: New feature
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification

### Checklist

- [x] Tests
  - [ ] **Regression tests**
  - [ ] C++ **Unit tests**
  - [x] **integration tests**
  - [ ] **resilience tests**
- [x] :book: CHANGELOG entry made
- [ ] :books: documentation written (release notes, API changes, ...)
- [x] Backports
  - [x] Backport for 3.10: https://github.com/arangodb/arangodb/pull/17652
  - [x] Backport for 3.9: https://github.com/arangodb/arangodb/pull/17653
  - [x] Backport for 3.8: https://github.com/arangodb/arangodb/pull/17654

#### Related Information

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket:
- [ ] Design document:

### CODE REVIEW

**Review by neunhoef — APPROVED** [review]
LGTM

File: arangod/GeneralServer/CommTask.cpp [hunk 1]
```diff
@@ -546,6 +546,11 @@ void CommTask::executeRequest(std::unique_ptr<GeneralRequest> request,
 // --SECTION-- statistics handling                             protected methods
 // -----------------------------------------------------------------------------
 
+void CommTask::setStatistics(uint64_t id, RequestStatistics::Item&& stat) {
+  std::lock_guard<std::mutex> guard(_statisticsMutex);
```
Comment (MBkkt) [inline arangod/GeneralServer/CommTask.cpp:550]:
```suggestion
  std::lock_guard guard{_statisticsMutex};
```

**Review by MBkkt — APPROVED** [review]

File: arangod/GeneralServer/H2CommTask.cpp [hunk 2]
```diff
@@ -366,40 +372,36 @@ void H2CommTask<T>::upgradeHttp1(std::unique_ptr<HttpRequest> req) {
...
+            // The HTTP/1.1 request that is sent prior to upgrade is assigned
+            // a stream identifier of 1 (see Section 5.1.1).
+            // Stream 1 is implicitly "half-closed" from the client toward the
+            // server
 
-        TRI_ASSERT(req->messageId() == 1);
-        auto* strm = me.createStream(1, std::move(req));
-        TRI_ASSERT(strm);
+            TRI_ASSERT(req->messageId() == 1);
+            auto* strm = me.createStream(1, std::move(req));
```
Comment (MBkkt) [inline arangod/GeneralServer/H2CommTask.cpp:396]: Btw is every http 1.1 connection will be blocked on single mutex here?
Is it really ok for us?

File: arangod/GeneralServer/H2CommTask.cpp [hunk 3]
```diff
@@ -53,6 +53,12 @@
 using namespace arangodb::basics;
 using std::string_view;
 
+namespace {
+constexpr std::string_view switchingProtocols(
+    "HTTP/1.1 101 Switching Protocols\r\nConnection: "
+    "Upgrade\r\nUpgrade: h2c\r\n\r\n");
```
Comment (mpoeter) [inline arangod/GeneralServer/H2CommTask.cpp:59]: `h2c` identifies the protocol where HTTP/2 is run over cleartext TCP. What about HTTP2 over TLS?

File: tests/test-definitions.txt [hunk 4]
```diff
@@ -42,6 +42,11 @@ shell_client priority=2500 parallelity=2 single suffix=vst -- --vst true
 shell_client_transaction priority=750 single suffix=http -- $EncryptionAtRest
 shell_client_transaction priority=1750 single suffix=http2 -- --http2 true
 shell_client_transaction priority=2750 single suffix=vst -- --vst true
+shell_client_traffic priority=250 parallelity=1 single suffix=http -- http true
+shell_client_traffic priority=250 parallelity=1 single suffix=http2 -- --http2 true 
+shell_client_traffic priority=250 parallelity=1 single suffix=vst -- --vst true
+shell_client_traffic priority=250 parallelity=1 single suffix=ssl-http -- http true --protocol ssl
+shell_client_traffic priority=250 parallelity=1 single suffix=ssl-http2 -- --http2 true --protocol ssl
```
Comment (mpoeter) [inline tests/test-definitions.txt:49]: Do we also have VST over SSL?

**Review by mpoeter — APPROVED** [review]

File: tests/test-definitions.txt (hunk 4 above)
Comment (jsteemann) [reply tests/test-definitions.txt:49]: I don't think that the combination of VST over SSL exists.

Comment (jsteemann) [reply tests/test-definitions.txt:49]: But actually... I may be wrong. Need to check.

Comment (jsteemann) [reply tests/test-definitions.txt:49]: Added that combination to tests as well, after fixing the testing framework for it.

File: arangod/GeneralServer/H2CommTask.cpp (hunk 3 above)
Comment (jsteemann) [reply arangod/GeneralServer/H2CommTask.cpp:59]: I won't make any functional additions in this PR, just fix the traffic accounting where possible.
I think HTTP/2 over TLS is negotiated using ALPN.

**Pull request merged**